_posts/2020-09-01-leveraging-rust-to-build-cross-platform-mobile-libraries.md

```diff
@@ -0,0 +1,79 @@
+
+I won't reiterate the full talk ([go watch it][video], really!), so this is just a brief overview of the Glean SDK itself.
+
+The Glean SDK is our approach to build a modern Telemetry library, used in Mozilla's mobile products and soon in Firefox on Desktop as well.
+
+The SDK consists of multiple components, spanning multiple programming languages for different implementations.
+All of the Glean SDK lives in the GitHub repository at [mozilla/glean](https://github.com/mozilla/glean).
+This is a rough diagram of the Glean SDK tech stack:
+
+![Glean SDK Stack](https://tmp.fnordig.de/blog/2020/glean-stack.png)
+
+On the very bottom we have `glean-core`, a pure Rust library that is the heart of the SDK.
+It's responsible for controlling the database, storing data and handling additional logic (e.g. assembling pings, clearing data, ..).
+As it is pure Rust we can rely on all Rust tooling for its development.
+We can write tests that `cargo test` picks up. We can generate the full API documentation thanks to `rustdoc`
+and we rely on `clippy` to tell us when our code is suboptimal.
+Working on `glean-core` should be possible by everyone that knows some Rust.
+
+On top of that sits `glean-ffi`.
+This is the FFI layer connecting `glean-core` with everything else.
+While glean-core is pure Rust, it doesn't actually provide the *nice API* we intend for users of Glean.
+That one is later implemented on top of it all.
+glean-ffi doesn't contain much logic.
+It's a translation between the proper Rust API of glean-core and C-compatible functions exposed into the dynamic library.
+In it we rely on the excellent [`ffi-support` crate](https://docs.rs/ffi-support/).
+ffi-support knows how to translate between Rust and C types, offers a nice (and safer) abstraction for C strings.
+glean-ffi holds *some* state: the instantiated global Glean object and metric objects.
+We don't need to pass pointers back and force. Instead we use opaque handles that index into a map held inside the FFI crate.
+
+The top layer of the Glean SDK are the different [language implementations](https://mozilla.github.io/glean/book/dev/core/internal/implementations.html).
+Language implementations expose a nice ergonomic API to initialize Glean and record metrics in the respective language.
+Additionally each implementation handles some special cases for the platform they are running on, like gathering application and platform data or hooking into system events.
+The nice API calls into the Glean SDK using the exposed FFI functions of `glean-ffi`.
+Unfortunately at the moment different language implementations carry different amounts of actual logic in them.
+Sometimes metric implementations require this (e.g. we rely on the clock source of Kotlin for timing metrics),
+in other parts we just didn't move the logic out of the implementations yet.
+We're actively working on moving logic into the Rust part where we can and might eventually use some code generation to unify the other parts.
```

Comment:
Maybe link to [Bug 1651382](https://bugzilla.mozilla.org/show_bug.cgi?id=1651382) here?